fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,683 @@
+        let num_vals = VIntU128::deserialize(reader)?.0 as u64;
+        let num_bits = u8::deserialize(reader)?;
+        let compact_space = CompactSpace::deserialize(reader)?;
+        let null_value = compact_space.unpack(null_value_compact_space);
+
+        Ok(Self {
+            null_value,
+            compact_space,
+            bit_unpacker: BitUnpacker::new(num_bits),
+            null_value_compact_space,
+            min_value,
+            max_value,
+            num_vals,
+            num_bits,
+        })
+    }
+}
+
+impl ColumnV2<u128> for CompactSpaceDecompressor {
+    fn get_val(&self, doc: u64) -> Option<u128> {
+        self.get(doc)
+    }
+
+    fn min_value(&self) -> u128 {
+        self.min_value()
+    }
+
+    fn max_value(&self) -> u128 {
+        self.max_value()
+    }
+
+    fn num_vals(&self) -> u64 {
+        self.params.num_vals
+    }
+
+    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = Option<u128>> + 'a> {
+        Box::new(self.iter())
+    }
+}
+
+impl ColumnV2Ext<u128> for CompactSpaceDecompressor {
+    fn get_between_vals(&self, range: RangeInclusive<u128>) -> Vec<u64> {
+        self.get_range(range)
+    }
+}
+
+impl CompactSpaceDecompressor {
+    pub fn open(data: OwnedBytes) -> io::Result<CompactSpaceDecompressor> {
+        let (data_slice, footer_len_bytes) = data.split_at(data.len() - 4);
+        let footer_len = u32::deserialize(&mut &footer_len_bytes[..])?;
+
+        let data_footer = &data_slice[data_slice.len() - footer_len as usize..];
+        let params = IPCodecParams::deserialize(&mut &data_footer[..])?;
+        let decompressor = CompactSpaceDecompressor { data, params };
+
+        Ok(decompressor)
+    }
+
+    /// Converting to compact space for the decompressor is more complex, since we may get values
+    /// which are outside the compact space. e.g. if we map
+    /// 1000 => 5
+    /// 2000 => 6
+    ///
+    /// and we want a mapping for 1005, there is no equivalent compact space. We instead return an
+    /// error with the index of the next range.
+    fn to_compact(&self, value: u128) -> Result<u64, usize> {
+        self.params.compact_space.to_compact(value)
+    }
+
+    fn compact_to_u128(&self, compact: u64) -> u128 {
+        self.params.compact_space.unpack(compact)
+    }
+
+    /// Comparing on compact space: 1.2 GElements/s
+    ///
+    /// Comparing on original space: .06 GElements/s (not completely optimized)
+    pub fn get_range(&self, mut range: RangeInclusive<u128>) -> Vec<u64> {
+        if range.start() > range.end() {
+            range = *range.end()..=*range.start();
+        }
+        let from_value = *range.start();
+        let to_value = *range.end();
+        assert!(to_value >= from_value);
+        let compact_from = self.to_compact(from_value);
+        let compact_to = self.to_compact(to_value);
+
+        // Quick return, if both ranges fall into the same non-mapped space, the range can't cover
```

Comment:
So we always map 0? What if values  go

```
|<  big blank space >|   ...a lot of values ...|< big blank space >|
```

Reply:
We don't map always 0, that's not what the comment says.

`|<  big blank space >|   ...a lot of values ...|< big blank space >|`

That's the actual data, but incoming requests can cover the whole space. If a range request is incoming for the same `big blank space`, we can early exit.

Reply:
I still don't understand... What does `self.to_compact` returning Err() means?
Doesn't it mean the value is not part of the covered value?
If so isn't the match on Err(_), Err(_) below wrong?

Reply:
Ah got it... I had missed pos1 == pos2 below